ext/odbc/php_odbc.c

```diff
@@ -812,30 +844,38 @@ void odbc_column_lengths(INTERNAL_FUNCTION_PARAMETERS, int type)
 /* {{{ Close all ODBC connections */
 PHP_FUNCTION(odbc_close_all)
 {
-	zval *p;
+	odbc_link *p;
+	zval *zv;
 
 	if (zend_parse_parameters_none() == FAILURE) {
 		RETURN_THROWS();
 	}
 
-	/* Loop through list and close all statements */
-	ZEND_HASH_FOREACH_VAL(&ODBCG(results), p) {
-		odbc_result_free((odbc_result*) p);
+	/* Loop through the connection list, now close all non-persistent connections and their results */
+	ZEND_HASH_FOREACH_PTR(&ODBCG(connections), p) {
+		if (p->connection) {
+			odbc_link_free(p);
```

Comment:
I changed the order: since connections will close their results as well, I started with this so that we can minimize iteration count of the result list below (only some dangling results may be still present after all the connections are closed, if any).